src/mlpack/tests/main_tests/gmm_train_test.cpp

```diff
@@ -0,0 +1,494 @@
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("gaussians", (int) 2);
+  SetInputParam("trials", (int) 0); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+// Checking that percentage is between 0 and 1.
+BOOST_AUTO_TEST_CASE(RefinedStartPercentageTest)
+{
+  arma::mat inputData(5, 10, arma::fill::randu);
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("refined_start", true);
+
+  Log::Fatal.ignoreInput = true;
+  SetInputParam("percentage", (double) 2.0); // Invalid
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+
+  SetInputParam("percentage", (double) -1.0); // Invalid
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+
+  Log::Fatal.ignoreInput = false;
+}
+
+// Samplings must be positive.
+BOOST_AUTO_TEST_CASE(GmmTrainSamplings)
+{
+  arma::mat inputData(5, 10, arma::fill::randu);
+
+  SetInputParam("input", std::move(inputData));
+  SetInputParam("refined_start", true);
+  SetInputParam("samplings", (int) 0); // Invalid
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+// Number of gaussians in the model trained from input model.
+BOOST_AUTO_TEST_CASE(GmmTrainNumberOfGaussian)
+{
+  arma::mat inputData(5, 10, arma::fill::randu);
+
+  SetInputParam("input", inputData);
+  SetInputParam("gaussians", (int) 2);
+
+  mlpackMain();
+
+  GMM* gmm = CLI::GetParam<GMM*>("output_model");
+
+  SetInputParam("input_model", gmm);
+
+  CLI::GetSingleton().Parameters()["input"].wasPassed = false;
+
+  SetInputParam("input", std::move(inputData));
+
+  mlpackMain();
+
+  GMM* gmm1 = CLI::GetParam<GMM*>("output_model");
+
+  BOOST_REQUIRE_EQUAL(gmm1->Gaussians(), (int) 2);
+}
+
+// Making sure that enabling no_force_positive doesn't crash.
+BOOST_AUTO_TEST_CASE(GmmTrainNoForcePositiveTest)
+{
+  arma::mat inputData(5, 10, arma::fill::randu);
+
+  SetInputParam("input", inputData);
+  SetInputParam("gaussians", (int) 1);
+  SetInputParam("no_force_positive", true);
+
+  mlpackMain();
+
+  GMM* gmm = CLI::GetParam<GMM*>("output_model");
+
+  SetInputParam("input_model", gmm);
+
+  CLI::GetSingleton().Parameters()["input"].wasPassed = false;
+
+  SetInputParam("input", std::move(inputData));
+
+  mlpackMain();
+
+  GMM* gmm1 = CLI::GetParam<GMM*>("output_model");
+
+  BOOST_REQUIRE_EQUAL(gmm1->Gaussians(), (int) 1);
+}
+
+// Ensure that Noise affects the final result.
+BOOST_AUTO_TEST_CASE(GmmTrainNoiseTest)
+{
+  arma::mat inputData;
+  if (!data::Load("data_3d_mixed.txt", inputData))
+    BOOST_FAIL("Unable to load train dataset data_3d_mixed.txt!");
+
+  SetInputParam("input", inputData);
+  SetInputParam("gaussians", (int) 2);
+  SetInputParam("noise", (double) 0.0);
+
+  size_t seed = std::time(NULL);
+  mlpack::math::randGen.seed((uint32_t) seed);
+  srand((unsigned int) seed);
+  arma::arma_rng::set_seed(seed);
```

Comment:
Now that we've verified that this seems to work reliably, I think we can remove this.  We can use `FixedRandomSeed()` instead.  I can make that change during merge.